src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,210 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/LdgIterator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Iterator for read-only device data. Use __ldg intrinsic to load data in
+ * read-only cache.
+ */
+template<class T>
+class LDGIterator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using difference_type = std::ptrdiff_t;
+    using value_type = T;
+    using pointer = std::add_pointer_t<T const>;
+    using reference = std::add_lvalue_reference_t<T const>;
+    using iterator_category = std::random_access_iterator_tag;
+    using self = LDGIterator<T>;
```

Comment:
Self or This, make it private, or just use LDGIterator because doesn't it default to using the same T?